deepinterpolation/generator_collection.py

```diff
@@ -897,112 +918,134 @@ class OphysGenerator(SequentialGenerator):
-            else:
-                self.raw_data = raw_data[:, :, :]
-            local_data = self.raw_data[0:average_nb_samples, :, :].flatten()
-        else:
-            local_data = raw_data[0:average_nb_samples, :, :].flatten()
-            self.cache_data = False
-
+        average_nb_samples = np.min(
+            [int(self.movie_data.shape[0]), self.movie_statistics_sample_size])
+        local_data = self.movie_data[:average_nb_samples].flatten()
         local_data = local_data.astype("float32")
 
-        self.local_mean = np.mean(local_data)
-        self.local_std = np.std(local_data)
+        self.local_mean = local_data.mean()
+        self.local_std = local_data.std()
+
+    @property
+    def movie_data(self) -> Union[np.ndarray, tf.Tensor]:
+        if self._movie_data is None:
+            with h5py.File(self.raw_data_file, "r") as movie_obj:
+                end_ind = self.end_frame+self.post_frame+self.pre_post_omission+1
+                total_frame_per_movie = movie_obj['data'].shape[0]
+                if self.end_frame > 0 and  total_frame_per_movie > end_ind:
+                    self._movie_data = movie_obj['data'][:end_ind]
+                else:
+                    self._movie_data = movie_obj['data'][()]
+            if self.gpu_cache_full:
+                logger.info("Caching full movie onto GPU")
+                self._data_tensor = tf.convert_to_tensor(
+                    self._movie_data, dtype="float")
+            if self.normalize_cache:
+                if not self.gpu_cache_full:
+                    self._movie_data = self._movie_data.astype("float32")
+                self._movie_data = _normalize(self._movie_data,
+                        self.local_mean,
+                        self.local_std)
+        return self._movie_data
+
+    def batch_tensor(self, index):
+        """Slices minimum movie required to generate batch for a given batch
+         index and caches it onto the GPU. If a previous batch is cached, a
+         new batch is generated by only copying the difference in frames.
+        """
+        start_ind = index*self.batch_size
+        end_ind = start_ind + self.batch_size + self.pre_frame \
+         + self.post_frame + 2*self.pre_post_omission - 2
+        
+        if self._cached_index == index-1:
+            batch_frames = tf.convert_to_tensor(
+                self.movie_data[end_ind-self.batch_size:end_ind], dtype="float")
+            if not self.normalize_cache:
+                batch_frames = _normalize(batch_frames, self.local_mean, self.local_std)
+            self._data_tensor = tf.concat(
+                [self._data_tensor[self.batch_size:], batch_frames], 0)
+            self._cached_index = index
         
-        if self.cache_data:
-            self.raw_data = ( self.raw_data.astype("float") \
-                - self.local_mean ) / self.local_std
+        elif self._cached_index == index:
+            return self._data_tensor
+        
+        else:
```

Comment:
clearer if you instead do 
```
if self._cached_index is not None:
    if self._cached_index == index-1:
        ....
    else:
        ...
else:
    ....
```

otherwise this else block could in theory be triggered unexpectedly

Reply:
The function of the first block is to copy just the difference of frames between the previous index, so this should only be run when self._cached_index == index-1. 

If there `self._cached_index is None or self._cached_index != index-1 or self._cached_index != index, it`, it should be caching the entire batch movie instead. 

So I think the current logic is correct.